javalib/src/main/scala/java/util/AbstractMap.scala

```diff
@@ -0,0 +1,209 @@
+package java.util
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+
+object AbstractMap {
+
+  private def equalsEntry[K, V](entry: Map.Entry[K, V], other: Any): Boolean = {
+    other match {
+      case entry2: Map.Entry[_, _] =>
+        (
+          (entry.getKey == null && entry2.getKey == null) ||
+          entry.getKey.equals(entry2.getKey)
+        ) && (
+            (entry.getValue == null && entry2.getValue == null) ||
+            entry.getValue.equals(entry2.getValue)
+          )
+      case _ => false
+    }
+  }
+
+  private def calcHashCode[K, V](entry: Map.Entry[K, V]): Int = {
+    val keyHash =
+      if (entry.getKey == null) 0
+      else entry.getKey.hashCode
+    val valueHash =
+      if (entry.getValue == null) 0
+      else entry.getValue.hashCode
+
+    keyHash ^ valueHash
+  }
+
+  class SimpleEntry[K, V](key: K, value: V) extends Map.Entry[K, V] with Serializable {
+    private var _key: K = key
+    private var _value: V = value
+
+    def this(entry: Map.Entry[K, V]) =
+      this(entry.getKey, entry.getValue)
+
+    override def equals(o: Any): Boolean =
+      equalsEntry(this, o)
+
+    def getKey(): K = _key
+
+    def getValue(): V = _value
+
+    override def hashCode(): Int = calcHashCode(this)
+
+    def setValue(value: V): V = {
+      _value = value
+      _value
```

Comment:
This returns the new `_value`, but it is supposed to return the *old* `_value`.
-> unit test